js/source/worker.js

```diff
@@ -153,7 +171,25 @@ util.extend(Worker.prototype, {
     'query features': function(params, callback) {
         var tile = this.loaded[params.source] && this.loaded[params.source][params.uid];
         if (tile) {
-            tile.featureTree.query(params, callback);
+
+            var id;
+
+            var classString = Object.keys(params.classes).join(' ');
+            if (this._cascadedClasses !== classString) {
+                this._cascadedClasses = classString;
+                for (id in this.styleLayersByID) {
+                    this.styleLayersByID[id].cascade(params.classes, {transition: false}, {});
+                }
+            }
+
+            if (this._recalculatedZoom !== params.zoom) {
+                this._recalculatedZoom = params.zoom;
+                for (id in this.styleLayersByID) {
+                    this.styleLayersByID[id].recalculate(params.zoom, { lastIntegerZoom: Infinity, lastIntegerZoomTime: 0, lastZoom: 0 });
+                }
+            }
+
+            callback(null, tile.featureTree.query(params, this.styleLayersByID));
         } else {
```

Comment:
@lucaswoj I just added these style layer creations and recalculations here as part of the style-property-aware featuresAt but I don't think they should stay here. Some of this is a duplication of what `Style` does but we don't need a full `Style` here. Any ideas on how this should be split out? Maybe a `StyleLayers`?